x/fees/keeper/integration_test.go

```diff
@@ -0,0 +1,658 @@
+			preBalance := s.app.BankKeeper.GetBalance(s.ctx, deployerAddress, denom)
+			gasTipCap := big.NewInt(10000)
+			gasFeeCap := new(big.Int).Add(s.app.FeeMarketKeeper.GetBaseFee(s.ctx), gasTipCap)
+			res := contractInteract(userKey, &contractAddress, nil, gasFeeCap, gasTipCap, &ethtypes.AccessList{})
+
+			developerCoins, validatorCoins := calculateFees(denom, params, res, gasFeeCap, 14)
+			feeColectorBalance := s.app.BankKeeper.GetBalance(s.ctx, feeCollectorAddr, denom)
+			balance := s.app.BankKeeper.GetBalance(s.ctx, deployerAddress, denom)
+			Expect(balance).To(Equal(preBalance.Add(developerCoins)))
+			Expect(feeColectorBalance).To(Equal(preFeeColectorBalance.Add(validatorCoins)))
+			s.Commit()
+		})
+
+		It("and fees are split 100-0 validators-developers, only validators receive fees", func() {
+			params = s.app.FeesKeeper.GetParams(s.ctx)
+			params.DeveloperShares = sdk.NewDec(0)
+			params.ValidatorShares = sdk.NewDec(1)
+			s.app.FeesKeeper.SetParams(s.ctx, params)
+
+			preFeeColectorBalance := s.app.BankKeeper.GetBalance(s.ctx, feeCollectorAddr, denom)
+			preBalance := s.app.BankKeeper.GetBalance(s.ctx, deployerAddress, denom)
+			gasTipCap := big.NewInt(10000)
+			gasFeeCap := new(big.Int).Add(s.app.FeeMarketKeeper.GetBaseFee(s.ctx), gasTipCap)
+			res := contractInteract(userKey, &contractAddress, nil, gasFeeCap, gasTipCap, &ethtypes.AccessList{})
+
+			_, validatorCoins := calculateFees(denom, params, res, gasFeeCap, 10)
+			feeColectorBalance := s.app.BankKeeper.GetBalance(s.ctx, feeCollectorAddr, denom)
+			balance := s.app.BankKeeper.GetBalance(s.ctx, deployerAddress, denom)
+			Expect(balance).To(Equal(preBalance))
+			Expect(feeColectorBalance).To(Equal(preFeeColectorBalance.Add(validatorCoins)))
+			s.Commit()
+		})
+
+		It("and fees are split 0-100 validators-developers, only developers receive fees", func() {
+			params = s.app.FeesKeeper.GetParams(s.ctx)
+			params.DeveloperShares = sdk.NewDec(1)
+			params.ValidatorShares = sdk.NewDec(0)
+			s.app.FeesKeeper.SetParams(s.ctx, params)
+
+			preFeeColectorBalance := s.app.BankKeeper.GetBalance(s.ctx, feeCollectorAddr, denom)
+			preBalance := s.app.BankKeeper.GetBalance(s.ctx, deployerAddress, denom)
+			gasTipCap := big.NewInt(10000)
+			gasFeeCap := new(big.Int).Add(s.app.FeeMarketKeeper.GetBaseFee(s.ctx), gasTipCap)
+			res := contractInteract(userKey, &contractAddress, nil, gasFeeCap, gasTipCap, &ethtypes.AccessList{})
+
+			developerCoins, _ := calculateFees(denom, params, res, gasFeeCap, 14)
+			feeColectorBalance := s.app.BankKeeper.GetBalance(s.ctx, feeCollectorAddr, denom)
+			balance := s.app.BankKeeper.GetBalance(s.ctx, deployerAddress, denom)
+			Expect(balance).To(Equal(preBalance.Add(developerCoins)))
+			Expect(feeColectorBalance).To(Equal(preFeeColectorBalance))
+			s.Commit()
+		})
+
+		It("we can update the withdrawal address if it is different than the deployer address", func() {
```

Comment:
why does the address have to be different?

Reply:
To avoid state bloat. If the addresses are the same on registration, we just don't store the withdrawal address. If they are the same on update, we fail the tx.